[Nicholascyx] ip

Nicholascyx opened the pull request:
# Kafka
> "The path to the future begins right here." - Kafka

Kafka aids you in following the path you are destined to be. It's,
- text-based
- easy to learn 
- ~~FUN~~ *SUPER* FUN to use

All you need to do is, 
1. download it from [here](https://github.com/Nicholascyx/ip/releases/tag/A-Release). 
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 👍 

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Here's the `main` method:
```ruby
public class KafkaLauncher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Nicholascyx committed:
Added Kafka.java, Added greet and exit message

Nicholascyx committed:
Added support for echo

Nicholascyx committed:
Added support for add and list function, added indentations

Nicholascyx committed:
Added new method for list

Nicholascyx committed:
Used ArrayList to store tasks

Nicholascyx committed:
Added [] for tasks and markAsDone() function

Nicholascyx committed:
Added support for Mark As Done

Nicholascyx committed:
Added Deadline, Event and Todo

Nicholascyx committed:
Added support for different events

Nicholascyx committed:
Added support for exceptions

Nicholascyx committed:
Added support for delete

Nicholascyx committed:
Added support for text-ui-test

Nicholascyx committed:
Add changes for text-ui-test

hm-yang1 commented on src/main/java/Kafka.java:
```diff
@@ -0,0 +1,136 @@
+                } else {
+                    if (userInput[0].equalsIgnoreCase("todo")) {
+                        if (userInput.length < 2) {
+                            throw new KafkaException("It seems you've left the details blank. Even the simplest tasks need some direction, don't you think?");
```
Exception message is quite long and used repeatedly. Maybe store the message as a String variable? Or write a function to throw a new exception?

hm-yang1 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,16 @@
+public class Event extends Task {
+
+    protected String from;
```
I don't see the protected access modifier often. Quite curious about the reason

hm-yang1 commented on src/main/java/Kafka.java:
```diff
@@ -0,0 +1,136 @@
+                  |     /  / _  | |_    _| | |/ /  / _  |
+                  |     \\ | |_| |   |  |   |   <  | |_| |
+                  |__|\\__\\ \\____|   |__|   |_|\\ \\  \\____|
+                """;
```
Nice. I wouldn't be able to create that and adds character to the bot

hm-yang1 commented on src/main/java/Kafka.java:
```diff
@@ -0,0 +1,136 @@
+        System.out.println("  Now you have " + this.tasks.size() + " task(s) in the list.");
+    }
+
+    public void createList() {
```
The function name is a little misleading as it does not create a new list. Perhaps getList or printList might be better?

hm-yang1 commented on src/main/java/Kafka.java:
```diff
@@ -0,0 +1,136 @@
+                    int taskNumber = Integer.parseInt(userInput[1]);
+                    kafka.delete(taskNumber);
+                } else {
+                    if (userInput[0].equalsIgnoreCase("todo")) {
```
Nested if statements are not as easy to read. Is it possible to just add more elif to handle the add task cases instead of nesting them in the else block?

hm-yang1 reviewed (commented):
Looks good.
Requested some cosmetic changes.

Nicholascyx committed:
Add support for saving tasks

Able to write the list of tasks into Kafka.txt

Nicholascyx committed:
Add a create new file function

Nicholascyx committed:
Add improvements for KafkaTextWriter

Easier format to read files

Nicholascyx committed:
Add support for loading tasks

It can now read from the Kafka.txt and update the list.

Nicholascyx committed:
Add datetime recogniton for deadlines

Exceptions to handle incorrect date format had been added as well.

Nicholascyx committed:
Add datetime recognition for events

Nicholascyx committed:
Add UI class functions

Some tasklist functions are also added, will continue in the next commit. Created files for Parser and Storage as well

Nicholascyx committed:
Add Storage class functions

Included some Parser and TaskList functions as well

Nicholascyx committed:
Add Parser class and TaskList class functions

Finished up all the functions and sync them together correctly

Nicholascyx committed:
Add all the classes into a package

Nicholascyx committed:
Merge branch 'add-gradle-support' of https://github.com/Nicholascyx/ip

Nicholascyx committed:
Add gradle into project

Nicholascyx committed:
Add JUnit tests

TaskList test is still not done yet, will complete it afterwards

Nicholascyx committed:
Improve JUnit test for TaskListTest

Nicholascyx committed:
Add Jar file

Nicholascyx committed:
Add JavaDoc comments

Nicholascyx committed:
Add Coding Standard changes

blackpanther9229 commented on src/main/java/kafka/Storage.java:
```diff
@@ -0,0 +1,89 @@
+        return tasks;
+    }
+
+    public void printFileContents(ArrayList<Task> tasks) {
```
You definitely need JavaDoc comment for this method

blackpanther9229 commented on src/main/java/kafka/Storage.java:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+
+    public void getNewFile(String filePath) {
```
This method should have JavaDoc comment

blackpanther9229 commented on src/main/java/kafka/Parser.java:
```diff
@@ -0,0 +1,93 @@
+
+public class Parser {
+
+    public static void parseCommand(String userInput, TaskList taskList, Storage storage, Ui ui) throws KafkaException, IOException, DateTimeParseException {
```
you need JavaDoc comment for this

blackpanther9229 commented on src/main/java/kafka/Parser.java:
```diff
@@ -0,0 +1,93 @@
+                ui.delete(taskToDelete, taskList);
+                break;
+            case "todo":
+                if (arguments.isEmpty()) {
+                    throw new KafkaException("It seems you've left the details blank. Even the simplest tasks need some direction, don't you think?");
+                }
+                Task todo = new Todo(arguments, false);
+                taskList.addTask(todo);
+                storage.writeToFile(taskList.tasks);
+                ui.addTask(todo, taskList);
+                break;
+            case "deadline":
+                if (arguments.isEmpty()) {
+                    throw new KafkaException("It seems you've left the details blank. Even the simplest tasks need some direction, don't you think?");
+                }
+                String[] deadlineParts = arguments.split("/by ");
+                if (deadlineParts.length < 2) {
+                    throw new KafkaException("It appears the details for this deadline task are off. Let's give it another go, shall we?");
+                }
+                LocalDateTime by = LocalDateTimeConverter.getLocalDateTime(deadlineParts[1]);
+                Task deadline = new Deadline(deadlineParts[0], by, false);
+                taskList.addTask(deadline);
+                storage.writeToFile(taskList.tasks);
+                ui.addTask(deadline, taskList);
+                break;
+            case "event":
+                if (arguments.isEmpty()) {
+                    throw new KafkaException("It seems you've left the details blank. Even the simplest tasks need some direction, don't you think?");
+                }
+                String[] eventParts = arguments.split("/from | /to ");
+                if (eventParts.length < 3) {
+                    throw new KafkaException("It appears the details for this event task are off. Let's give it another go, shall we?");
+                }
+                LocalDateTime from = LocalDateTimeConverter.getLocalDateTime(eventParts[1]);
+                LocalDateTime to = LocalDateTimeConverter.getLocalDateTime(eventParts[2]);
+                Task event = new Event(eventParts[0], from, to, false);
+                taskList.addTask(event);
+                storage.writeToFile(taskList.tasks);
+                ui.addTask(event, taskList);
+                break;
```
It is better to add some comments for clarifying

Oceankoh commented on src/main/java/kafka/TaskList.java:
```diff
@@ -0,0 +1,51 @@
+        if (tasks.isEmpty()) {
+            return;
+        }
+        String listMessage = "";
```
Minor nitpick, but you could consider using StringBuilder for some improvement in time complexity. It is really minor though as the strings are probably not long enough to result in any real world performance benefits :)

Oceankoh commented on src/main/java/kafka/Parser.java:
```diff
@@ -0,0 +1,93 @@
+        String command = splitInput[0].toLowerCase();
+        String arguments = splitInput.length > 1 ? splitInput[1] : "";
+
+        switch (command) {
```
Wow! This is a really long switch case. Could this be broken up into smaller parts? Maybe try creating a method for each case! This would improve code readability!

Oceankoh commented on src/test/java/seedu/kafka/StorageTest.java:
```diff
@@ -0,0 +1,65 @@
+package seedu.kafka;
+
+import kafka.Deadline;
+import kafka.Storage;
+import kafka.Todo;
+import kafka.Event;
+import kafka.Task;
+import org.junit.jupiter.api.Test;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Does this follow the package import coding standards?

Oceankoh commented on src/main/java/kafka/Kafka.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    public static void main(String[] args) {
+        new Kafka("C:/Users/Nicholas/Downloads/Kafka.txt").run();
```
Would this work if I run it on a non-windows PC?

Elinengu replied:
Hello, I think it might not work on other computer (even if it's windows) because the path directory won't be the same. Perhaps it's better to use a relative path instead? And maybe it's better to handle the case where a folder or a txt file has not being created yet to prevent checked exception?
![Screenshot 2024-09-07 191711](https://github.com/user-attachments/assets/e0931b04-4216-40b4-abf1-14a0354e8a81)

Oceankoh reviewed (commented):
Great Effort! Loved how you customized the different print statements

Elinengu commented on src/test/java/seedu/kafka/StorageTest.java:
```diff
@@ -0,0 +1,65 @@
+public class StorageTest {
+
+    @Test
+    public void testWriteTodoToFile() throws IOException {
```
I think this might not be the naming convention for testing method. Perhaps can double check it on topics?

Elinengu commented on src/main/java/kafka/Storage.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    public void writeToFile(ArrayList<Task> tasks) {
+        try (FileWriter fw = new FileWriter(filePath)) {
+            for (Task t : tasks) {
+                if (t instanceof Todo todo) {
+                    fw.write("T | " + todo.isDone + " | "
+                            + todo.description + System.lineSeparator());
+                } else if (t instanceof Deadline deadline) {
+                    fw.write("D | " + deadline.isDone + " | "
+                            + deadline.description + " | " + deadline.by + System.lineSeparator());
+                } else if (t instanceof Event event) {
+                    fw.write("E | " + event.isDone + " | "
+                            + event.description + " | " + event.from + " | " + event.to
+                            + System.lineSeparator());
+                }
+            }
```
I'm not quite sure about it (since I couldn't run the file) but I think this method might cause only 1 line of task to be written in the final txt file because the writer object is not being created in append mode. Perhaps it's better to check it out?

Nicholascyx committed:
Add find function for the tasks

Nicholascyx committed:
Merge branch 'branch-A-JavaDoc'

Nicholascyx committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/kafka/Parser.java

Nicholascyx committed:
Add JavaDoc comments for find functions and resolve conflicts

Nicholascyx committed:
Add support for GUI

Some functions are not yet finished

Nicholascyx committed:
Add more support for GUI

Fix some issues on the chatbot not printing the list of tasks

Nicholascyx committed:
Add some code quality changes

Parser method has over 30 lines of code which violates code quality.

Let's make each switch case into a different function so that there's only one line of code under every switch case.

Nicholascyx committed:
Add some additional changes for parser command

Nicholascyx committed:
Add assertions

Some parts of the code need important assumptions that should hold.

Let's add some code using assert function in java to document assumptions.

Nicholascyx committed:
Change more codequality relevant stuff

Some part of my code violates the code quality standards.

Let's change them and make it more readable and understandable.

Nicholascyx committed:
Add small adjustments to throw IOException

Nicholascyx committed:
Merge pull request #2 from Nicholascyx/branch-A-Assertions

Add assertions

Nicholascyx committed:
Merge branch 'master' into branch-A-CodeQuality

Nicholascyx committed:
Revert "Merge branch 'master' into branch-A-CodeQuality"

This reverts commit 2c9400ff9826fd76378c9940ea9699278a29de18, reversing
changes made to 16a76a575932d412a098ecaadc88e841dbc6fd00.

Nicholascyx committed:
Merge pull request #3 from Nicholascyx/branch-A-CodeQuality

Merge code quality branch

Nicholascyx committed:
Add more JavaDocs

Nicholascyx committed:
Add snooze function

Users sometimes may face the problem of wanting to change the deadline or event date, or delay an event.

Let's add a snooze function for users to edit their deadline and event dates.

Nicholascyx committed:
Add support for assertions

Made some mistakes in the assertion branch.

Let's fix it by adding it again.

Nicholascyx committed:
Merge branch 'branch-B-Snooze'

Nicholascyx committed:
Update the filepath

Ensures it can be opened across different platforms.

Nicholascyx committed:
Create jar file

Nicholascyx committed:
Add Readme

Nicholascyx committed:
Update README.md
